### PR TITLE
[ADD] osi_fifo_serialized_fix

### PR DESCRIPTION
**Pull request by max3903**

### CODE REVIEW

**Comment by b-kannan** [comment]
@osi-scampbell @agyamuta The stock_quant.py change (last commit) only considers lot and not lot. What about serial number quants?

**Comment by osi-scampbell** [comment]
> @osi-scampbell @agyamuta The stock_quant.py change (last commit) only considers lot and not lot. What about serial number quants?

@b-kannan I think I got tunnel vision and only wrote a cases for Lot and SN. The 'else' in this case is supposed to be for SN tracked objects, however I believe this would affect products with no tracking which is not intended. The commit I just pushed alters the method to only check for Lot and SN, if it is not tracked, then we will use core odoo workflow.

**Comment by b-kannan** [comment]
> > @osi-scampbell @agyamuta The stock_quant.py change (last commit) only considers lot and not lot. What about serial number quants?
> 
> @b-kannan I think I got tunnel vision and only wrote a cases for Lot and SN. The 'else' in this case is supposed to be for SN tracked objects, however I believe this would affect products with no tracking which is not intended. The commit I just pushed alters the method to only check for Lot and SN, if it is not tracked, then we will use core odoo workflow.

Thank you.

File: osi_fifo_serialized_fix/models/stock_quant.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+            if quant.product_id.cost_method == 'fifo':
+                svl_ids = self.env['stock.valuation.layer'].\
+                    search([('product_id', '=', quant.product_id.id),
+                            ('value', '>', 0)])
+                for svl_id in svl_ids:
+                    if quant.lot_id in svl_id.lot_ids:
+                        if quant.product_id.tracking == 'lot':
+                            quant.value = svl_id.value / svl_id.quantity \
+                                * quant.inventory_quantity * -1
+                        else if quant.product_id.tracking == 'serial':
```
Comment (max3903) [inline osi_fifo_serialized_fix/models/stock_quant.py]:
```suggestion
                        elif quant.product_id.tracking == 'serial':
```

File: osi_fifo_serialized_fix/views/stock_valuation_layer.xml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+
+<odoo>
```
Comment (max3903) [inline osi_fifo_serialized_fix/views/stock_valuation_layer.xml]:
```suggestion
<odoo>
```

File: osi_fifo_serialized_fix/models/stock_move.py [hunk 3]
```diff
@@ -31,67 +31,68 @@ def _action_done(self, cancel_backorder=False):
...
+                            val = move.product_id.\
+                                _prepare_out_svl_vals(svl_id[1], self.env.user.company_id)
+                            val_obj = self.env['stock.valuation.layer'].browse(svl_id[0])
+                            val['unit_cost'] = val_obj.unit_cost
+                            val['company_id'] = self.env.user.company_id.id
+                            val['lot_ids'] = [(6, 0, svl_id[2])]
+                            val['account_move_id'] = ji_ids[0].move_id.id
+                            # The qty of serial products is always 1, lots could be > 1
+                            if line_id.product_id.tracking == 'lot':
+                                val['quantity'] = line_id.qty_done * -1
```
Comment (sorrytosay) [inline osi_fifo_serialized_fix/models/stock_move.py:50]: For outgoing

File: osi_fifo_serialized_fix/models/stock_move.py [hunk 4]
```diff
@@ -31,67 +31,68 @@ def _action_done(self, cancel_backorder=False):
...
+                            val['value'] = val_obj.unit_cost * val['quantity'] * -1
```
Comment (sorrytosay) [inline osi_fifo_serialized_fix/models/stock_move.py:51]: For outgoing transfesr, the 'value' in SVL layer should be negative. It becomes positive. value['value'] = val_obj.unit_cost * val['quantity'] *  -1

**Review by sorrytosay — CHANGES_REQUESTED** [review]
Disregard this changes

**Review by patrickrwilson — APPROVED** [review]
This is an old PR but used in some projects for a while now. This looks good to me so we should move it forward.

**Pull request merged**